### DISCUSSION TITLE
Awesome-i3?

### DISCUSSION THREAD

**Discussion by budRich** [i3/i3#4509] [Ideas]
Nice to see you choosing a more appropriate platform for i3 support/questions/discussions

I have had this idea for quite some time, but never made it reality.

# Awesomei3

I am not talking about the awesome awesomewm, but this nice GitHub meme that is known as **Awesome Lists** (https://github.com/sindresorhus/awesome/)  

There exist a lot of third-party i3 *extensions* or what to call them. And it would just be neat to collect links to all of them (most are already on GitHub) and setup a Awesome-i3 repo with those links. That repo could also have the **Discussions** feature enabled.  

I think it would benefit this, official i3 support *forum*, if there existed an *unofficial* one. Where questions about such things could be asked, and a more natural place where we can advertise our own i3hacks and config tweaks. 

I think I have a list of i3 related links (these include both tools on github, but also blogposts and stuff like that). And I could setup and host the repo on my personal github account or budlabs. But maybe the most appropriate would be if it would be i3/awesome?

**Comment by Airblader** [comment]
Thanks for the idea! In general I think a curated list of resources makes a lot of sense.

How do 

> I think it would benefit this, official i3 support forum, if there existed an unofficial one

and 

> But maybe the most appropriate would be if it would be i3/awesome?

go together? If it's hosted in the i3 organization, it kind of automatically becomes an official resource.

**Reply by budRich** [reply]
It will benefit this support forum, if there exist a place to redirect questions and such that is related to third party i3 tools. It will make this place leaner, and i think, less frustrating to moderate. 

It doesn't mean it couldn't be a repository under the i3 organization. An awesome repo is really just a markdown file with links, so no unofficial software will be published in that sense under the i3 organization. But it might be of interest to have some sort of control over the repository for the i3 organization. If f.i. the maintainer of the repo lose interest. Even if it is not part of i3 officially it would look better to either keep the awesome list alive (assigning a new maintainer) or shutting it down. That might not be possible to do, if one person is holding the keys. If "discussions" is enabled on the repository, it might be nice to be able to, you know, have a say in what can be discussed there.

At the same time, it could and probably would work completely fine outside the i3 organisation. And it is possible to move (change ownership) of a repository after it is created (probably only works with the "git" stuff, issues, and discussions are not moved afaik).

---

I once created a organisation, i3script, with this kind of thing in mind, maybe that is still the best place for the awesome-i3 list. And then I can try to find at least one more co-maintainer, to counter the bus factor.

**Reply by budRich** [reply]
i checked, and i actually already had a (private) i3script/awesome-i3 repo setup. I made it public now, but i know i have a lot more links stored on my local machine i can add, and many many more to add.

https://github.com/i3script/awesome-i3

**Comment by Airblader** [comment]
Not losing ownership should it be orphaned is certainly a good argument, but again an argument for it being official — if it isn't official, the i3 organization has no reason to have control over it. :-)

That said, I'm not against having it be (let's say semi-)official. But it requires adding someone to the organization, and that's something that requires a certain amount of trust and commitment. Those qualities are earned over time (but this is just my personal IMO). 

A compromise could be to maintain it as a thread on Discussions for now and we pin that thread for visibility. If we ever move it into a repository afterwards, we just edit the thread to link to that repository. Do you think this would be a good way to get started? Also, does this mean you'd be willing to take ownership on this list?

**Reply by budRich** [reply]
My intention with this thread was not to squeeze myself into the i3 organization, i just thought of it now, that maybe it is better if it was i3/awesome-i3. So the "compromise" is great, and i can take ownership of the list. And lets do it as a thread here now, i make the repo i had unlisted again. And we see how it works, i am more then fine if i end up not being  the maintainer/moderator of a future official/unofficial repo.

**Comment by budRich** [comment]
is the thread this thread?
Or a new one in "idea" called awesome-i3 - curated list of i3 related stuff?

If this is the thread, and if it is possible to pin posts to the top of the thread i will post the list i had at the repo in the next post, so it can be pinned to the this thread :)

**Comment by budRich** [comment]
# awesome-i3 - a curated list of all awesome i3 related stuff

- i3 fetchers
  - [i3-match]
  - i3get
  - [py3tree], [i3tree]
  - [i3help]
  - [i3-cheat]
- ipc wrappers
- config tools
- i3status generators
  - [i3-dstatus]
  - [i3status-rs]  
  - [goi3bar]
  - [bumblebee-status]
- layout manipulators
  - [i3ipc-dynamic-tiling]
  - [i3-balance-workspace]
  - [i3-resurrect]
- wm management
  - [i3-scratchpad]
  - [dopaminie-2000], [i3-parkinson], [dopamine]
  - [i3-wk-switch]
  - [i3gopher]
- external programs
  - compositor hacks
    - [invert window colors picom xprop]
    - [i3-workspace-brightness]
- articles
  - [xfce4-i3-arch]
- dotfiles
  - [Fabian Untermoser - i3 dotfiles and scripts]
- theming
  - [i3-style]
- old stuff
  - [i3-py]
  
    
  
   
  
  
  
  
  
[i3tree]: https://github.com/sciurus/i3tree
[i3-scratchpad]: https://gitlab.com/aquator/i3-scratchpad
[i3help]: https://github.com/glennular/i3help
[Fabian Untermoser - i3 dotfiles and scripts]: https://gitlab.com/FabianUntermoser/dot-files/-/tree/master/i3
[dopaminie-2000]: https://github.com/EllaTheCat/dopamine-2020
[i3-parkinson]: https://github.com/EllaTheCat/i3-parkinson
[dopamine]: https://github.com/EllaTheCat/dopamine
[i3-wk-switch]: https://github.com/tmfink/i3-wk-switch
[invert window colors picom xprop]: https://old.reddit.com/r/i3wm/comments/kbw3a5/shortcut_for_inverting_a_windows_colors/
[i3-dstatus]: https://github.com/altdesktop/i3-dstatus
[i3ipc-dynamic-tiling]: https://github.com/chlyz/i3ipc-dynamic-tiling
[i3-workspace-brightness]: https://github.com/orhun/i3-workspace-brightness
[py3tree]: https://github.com/DavidAntliff/py3tree
[i3status-rs]: https://github.com/greshake/i3status-rust
[goi3bar]: https://github.com/denbeigh2000/goi3bar
[i3-match]: https://gitlab.com/ferreum/i3-match
[i3gopher]: https://github.com/quite/i3gopher
[i3-style]: https://github.com/altdesktop/i3-style
[i3-cheat]: https://github.com/TheCynicalTeam/i3-cheat
[i3-volume]: https://github.com/hastinbe/i3-volume
[i3-resurrect]: https://github.com/JonnyHaystack/i3-resurrect
[i3-balance-workspace]: https://github.com/atreyasha/i3-balance-workspace
[i3-py]: https://github.com/ziberna/i3-py
[bumblebee-status]: https://github.com/tobi-wan-kenobi/bumblebee-status
[azote]: https://github.com/nwg-piotr/azote
[xfce4-i3-arch]: https://github.com/EnricoVogt/xfce4-i3-arch

**Reply by RayZ0rr** [reply]
Hey there @budRich. Any updates on this? I think this should be made a separate repo apart from official i3, as it is community maintained. Also, a mention about it can be made in official repo. You can add other collaborators/maintainers so that it would become stale when you are busy, just like other awesome lists

**Comment by Airblader** [comment]
I meant to post it as a new separate thread, I can then pin that thread to the top for the entire Discussions page. :-)

I also didn't mean to say anyone is trying to squeeze themselves into an organization. I think it's awesome that you want to do this, and thank you for that. I just meant to say that people get added to the organization after having shown some level of commitment/contribution over time since being part of the organization comes with certain rights. I hope that makes it a bit more clear.

**Reply by budRich** [reply]
>I think it's awesome that you want to do this

Does this mean I can add myself to the list?

**Reply by budRich** [reply]
i will treat this thread as the thread till the thread is created.  

The list should imo comply with the [Awesome manifesto](https://github.com/sindresorhus/awesome/blob/main/awesome.md)

My current list doesn't because:
- [ ] No table of contents
- [ ] No comments
- [ ] Some of the things might not be awesome (f.i. old stuff)
- [ ] no badge
- [ ] no list description

**Reply by Airblader** [reply]
Yeah, of course you can add yourself to it. Thanks for doing this!

**Reply by totoro-ghost** [reply]
Yes I was thinking of making this too, 

add theses to the list too

- [i3expo](https://github.com/morrolinux/i3expo-ng)
- [i3 theme maker online](https://thomashunter.name/i3-configurator/)

also i have some scripts, i'll make a repo of them if you make this (awesome/i3) a repo.

**Reply by budRich** [reply]
i3expo, cool I hadn't seen it before. But taking a screenshot on **every** window event? Sounds like it will not be that awesome if you for instance have a timer (muisic/video player) in the titlebar. @totoro-ghost have you tried it yourself? Theme maker is awesome classics.

**Comment by budRich** [comment]
I will experiment a bit with the formatting of the list. It doesn't have to be a bullet list, and it kind of looks off if we add comments and images to the links.  Maybe this format is better:

(*workspace enhancements would be a category, here with two entries*)

---

### workspace enhancements

Tools that will change or extend the workspace functionality. 
<details>
  <summary><b>i3expo</b> - overview of all workspaces with i3 </summary>

  <br>
  
  - **author** - [morrolinux.it](https://morrolinux.it) 
  - **source** - [GitHub](https://github.com/morrolinux/i3expo-ng)
  - **language** - python, C
  - **license** - MIT
  - **packages** - [AUR](https://aur.archlinux.org/packages/i3expo/)

  uses the pygame framework to create a interactive workspace
  overview. Takes a screenshot every time the i3 IPC reports a window event.  

  ![i3expo screenshot](https://github.com/morrolinux/i3expo-ng/raw/main/img/ui.png)  
</details>


<details>
  <summary><b>i3-workspace-brightness</b> - individual brightness levels for each workspace </summary>

  <br>
  
  - **author** - [orhun](https://github.com/orhun/) 
  - **source** - [GitHub](https://github.com/orhun/i3-workspace-brightness)
  - **language** - rust
  - **license** - GPL-3.0
  - **packages** - [AUR](https://aur.archlinux.org/packages/i3-workspace-brightness/)

  allows individual brightness levels for each workspace and changes the brightness
  automatically while switching between them.  
</details>

[i3expo]: https://github.com/morrolinux/i3expo-ng

**Comment by budRich** [comment]
In the [awesome manifesto](https://github.com/sindresorhus/awesome/blob/main/awesome.md) they say that  
>You should rather leave stuff out than include too much  

And I agree with that. If this list will become a long lasting thing it will be necessary to remove things that f.i. doesn't work with the latest stable i3 release and such. So I thought it will probably be a good idea to also add to the *summary*:  

- **last testet on i3 version** - 4.19

And if a entry on the list doesn't get tested (and of course pass the test) with the latest i3 release it should get removed. This should imo also apply to guides (blogposts, videos e.t.c). If something in a guide is obsolete, but the guide still can be considered awesome, issues like that should be clearly stated in the list.

**Reply by budRich** [reply]
This doesn't mean that the maintainer of the list have to test every single entry for each release. I am more thinking that those who use the awesome stuff will report if they are still awesome. And lets say i3 v4.20 is released, all entries on the list that hasn't passed the 4.19 test should be removed. Off course entries that fall out can be re-added.

It also doesn't have to happen immediately with a new release, could be a months time to comply or something.

---

with all that said, i personally would like some kind of index of **all** external i3 resources, be them awesome or working is not really the point. I might set something like that up, maybe..

**Comment by eater** [comment]
Awesome! I've never heard of awesome lists, but I started [this list](https://github.com/noodlor/i3-resources) once. It never made it off the back burner because I have no time ever. 

I would contribute to a repo. I also like the way https://vimawesome.com/ does it.

**Reply by budRich** [reply]
wow that vimawesome looked very neat, I have a halfbaked hugo project that could turn into something similar. Hugo is really good when it comes to taxonomies (tags and stuff). I will see if the you have any awesome thing on your list and add them. i just glanced over and saw some things that are awesome, but not really i3-awesome, like polybar. I think it will be very good to not include things that aren't explicitly for i3 in the list. Yeah, i will see if I can draft something like a criteria for what could be included and not.

**Comment by ghost** [comment]
Actually, I'm working on i3db.org right now :-) It will have similar idea like i3 awesome list but much better ^_^ I know I'm running late on this project, just need to finish few other things before I continue on i3b.org development. But I haven't abandoned it. So stay tuned! Sincerly, @tuxy @i3db (https://github.com/i3db)

**Reply by budRich** [reply]
I just needed to read the first section i3db/scripts/bash.md to understand it is a top notch project ;)

Seriously though, it looks really cool and it is more of what i mentioned regarding a collection/database/archive of **all** i3 related stuff, and i think both the awesome list and i3db can live side by side in harmony. 

The awesome-list being curated with only the best, updated and tried things, while i3db (who of course needs to be included in the awesome list) could try to include everything?

One thing I have been thinking of is that it would be cool to crawl f.i. the subreddit and gist for nice code blocks and save and archive them as well. I guess it is somewhat grayzone license wise, but in most such cases it will probably not be impossible to get in contact with the author and get their approval for republishing.

I can see if i can get that hugo project i was working on in shape, then you could use that for i3db webpage if you want to, you can still use JSON as the "input" to generate. The nice thing with hugo is that it will automatically create tags and categories pages, and its all static . Projects and list like this often get "too" big without a good tag system, so that one item can be found under multiple categories and or tags.

**Reply by ghost** [reply]
> I just needed to read the first section i3db/scripts/bash.md to understand it is a top notch project ;)

Keep laughing, LOL @ https://github.com/i3db/scripts/blob/main/bash.md

> Seriously though, it looks really cool and it is more of what i mentioned regarding a collection/database/archive of all i3 related stuff, and i think both the awesome list and i3db can live side by side in harmony.

At first I wanted to create awesome lists, but than decided against it. I will create something much more useful. It will be very organized repository of not only i3db resources but also many other linux related software. But it will take some time for me to even start the project, maybe in 2022. I'm currently working on finishing v2 of my TVguide project. But you are welcome create this awesome lists, this will only help when the time comes.

> One thing I have been thinking of is that it would be cool to crawl f.i. the subreddit and gist for nice code blocks and save and archive them as well. I guess it is somewhat grayzone license wise, but in most such cases it will probably not be impossible to get in contact with the author and get their approval for republishing.

I'm against any automation, because this will just create too much content that you will have to sort throught. This is not the way I want i3db to be. I want to maintaine it manually, just like I do my TVguide project. I know its a lot of work, but in the end it will be much more useful to everyone. And yes of course I will have removal/aproval features on site.

> I can see if i can get that hugo project i was working on in shape, then you could use that for i3db webpage if you want to, you can still use JSON as the "input" to generate. The nice thing with hugo is that it will automatically create tags and categories pages, and its all static . Projects and list like this often get "too" big without a good tag system, so that one item can be found under multiple categories and or tags.

No thank you. I won't be using Hugo for this, I will code my own site from scratch and yes it will use JSON for data. So it will be very easy to request addition/removal/aproval etc. And i3db will also have cagegories, tags and much more.

**Reply by budRich** [reply]
> > One thing I have been thinking of is that it would be cool to crawl f.i. the subreddit and gist for nice code blocks...
> 
> I'm against any automation...

I was actually thinking of crawling/archiving code blocks manually, but automating parts of it is probably a good idea. If anyone is reading this and know a way to make files of all (unique) codeblocks (that starts with a shebang) named as the comment-id from r/i3wm , i am intested. Also scrubbing for links to github and gist.github would be of interest.

**Reply by ghost** [reply]
@budRich I know how to do this, but I won't do it for i3db project. However, if you want to hire me on hourly or per-project basis I can help with this scrapers. Don't have the time to do this web scraping it for free sorry. Or see if anyone else is interested.

**Comment by budRich** [comment]
I used curl, sed and the [pushshift.io API](https://github.com/pushshift/api) to collect links matching the query "github":  
https://gist.github.com/budRich/23f8165ce1821edd4dc27f9fc3f820d6

There are lots of none github links in the end result, because i just grabbed all links from the results, but i cleaned it up a bit, its sorted and no duplication.

I will make some more queries (pastebin, gitlab, gist e.t.c) some other time, and add those results to the gist linked above. Next step is to sort out all broken links. 

below is my hackish bash script that helped me do this (i manually cleaned up the results afterwards and then did a `sort -u` of the file):  
*NB there are probably much better ways of doing it. I did the day-based looping because i wasn't sure how many results i would get and if i messed up i could just start from the last value of days.* :sunglasses:  

``` bash
#!/bin/bash

search_for="github.com"
size=100 # max 500
sub=i3wm # subreddit
output=~/urls-from-"$sub"

# https://github.com/pushshift/api

API_comment='https://api.pushshift.io/reddit/search/submission/?'
# API_comment='https://api.pushshift.io/reddit/search/comment/?'
query="q=$search_for&subreddit=$sub&size=$5&fields=selftext,title"
# query="q=$search_for&subreddit=$sub&size=$5&fields=body"

# extract markdown urls
# sed -rn 's@^.*\[(.*)\]\((.*)/.*$@\2@p'

[[ -f $output ]] && rm -f "$output"
mkdir -p "${output%/*}"

days=-99
while true ; do
  ((days+=100))
  echo "$days"
  json=$(curl "$API_comment${query}&before=${days}d&after=$((days+100))d")
  [[ $json ]] || break
  <<< "$json" sed -rn 's@^.*\[(.*)\]\((.*)/.*$@\2@p' >> "$output"
  ((days > 3000)) && break
  sleep .2
done
```

**Reply by budRich** [reply]
i just realized that the regex i used to grab links is not that good. i will try this next time: `^.*\[(.*)\]\(([^)]+)/.*$` , should fix all the jank output. maybe

**Reply by ghost** [reply]
I recommend you using python or ruby for web scraping, you can do much cleaner and smarter code with that. And the way you scraping now you just creating too much double work for yourself. There will be too many unrelated links.

**Reply by budRich** [reply]
thanks. I learned that making one regex to find all urls in an arbitrary text is more or less impossible. I found a Go program, `xurls` that worked fine, so i use that now. I also use `jq -r` to spit out the raw markdown. This however still leaves escaped underscores (because some markdown think underscore starts bold (what a stupid syntax)). So they need to be removed as well. But these combined grabbed me a lot more urls. I am fine grabbing the unrelated urls, i rather get too many then too few. It is quite easy to spot most of the jank ones, and when the links are sorted alphabetically, one can just select a block of them and delete.  

I made (copied from so) one more script to check for 404 and redirects and removed duplicates, all automated it takes a minute or two to search the whole subreddit (all comments and posts) with a search term and grab the links.

I am kind of surprised i didn't find more stuff with this. If I search for **i3** in the AUR i will actually find more awesome stuff then i did with this (excluding pastebin). Also only ~10 unique gitlab links. 0 sr.ht (excluding https://sr.ht). 

PS. I also tried a subredditscraper python script i found on github (will not link), but i am pretty sure it was a disguised malware and i think i will need to audit/re-install my system because i ran it (i read the source before executing, but it did weired things anyways, disguised. I should have known looking at the ascii dogecoin banner...)...

``` bash
#!/bin/bash

search_for=(github gitlab pastebin sr.ht youtube)
size=100 # max 500
sub=i3wm # subreddit

for query in "${search_for[@]}" ; do

  output=~/git/bud/src/new/reddit-get/"${query}"

  days=-99
  API_submission="https://api.pushshift.io/reddit/search/submission/?subreddit=$sub&size=$5&fields=selftext,title&q=$query"
  API_comment="https://api.pushshift.io/reddit/search/comment/?subreddit=$sub&size=$5&fields=body&q=$query"

  while true ; do
    db=$((days+=100)) da=$((db+100))
    echo "Q: $query - D: $days"
    {
      curl "$API_comment&before=${db}d&after=${da}d" | jq -r '.data | .[] | .body'
      sleep 0.1
      curl "$API_submission&before=${db}d&after=${da}d" | jq -r '.data | .[] | .selftext , .title'
      sleep 0.1
    }  | tr -d \\ | xurls >> "${output}-unsorted"
    ((days > 2500)) && break
  done
  sort -u "${output}-unsorted" > "${output}-sorted"
  cat "${output}-sorted" >> ~/git/bud/src/new/reddit-get/ALL-unsorted
done

sort -u ~/git/bud/src/new/reddit-get/ALL-unsorted > ~/git/bud/src/new/reddit-get/ALL-sorted
```


But what is useful doing this is that i get a register of a lot of *normal* i3 users github profiles, and it is in their `dotfiles/bin` directories where some cool stuff can be found.

**Reply by ghost** [reply]
Fine scrape away than, I won't be waisting time on i3db than :-) Have fun sorting!

**Reply by budRich** [reply]
whatever..

**Comment by budRich** [comment]
Sorry, that i faded away from this and that. Just saw that github now has "star lists", its better than nothing. I will try to get back to this project soon(tm)

[i3 goodies](https://github.com/stars/budRich/lists/window-i3-goodies)

**Comment by Syphdias** [comment]
I made another list in case someone is interested: https://github.com/Syphdias/awesomeish-i3

**Reply by budRich** [reply]
Nice @Syphdias but not enough https://github.com/budlabs/i3ass ;)
